tests/components/bmw_connected_drive/test_button.py

```diff
@@ -86,7 +79,100 @@ async def test_update_failed(
             "button",
             "press",
             blocking=True,
-            target={"entity_id": "button.i4_edrive40_flash_lights"},
+            target={"entity_id": "button.i4_edrive40_activate_air_conditioning"},
         )
-    assert RemoteServices.trigger_remote_service.call_count == 1
-    assert BMWDataUpdateCoordinator.async_update_listeners.call_count == 0
+    assert hass.states.get(entity_id).state == old_value
+
+
+@pytest.mark.parametrize(
+    ("entity_id", "state_entity_id", "new_value", "old_value"),
+    [
+        (
+            "button.i4_edrive40_activate_air_conditioning",
+            "switch.i4_edrive40_climate",
+            "on",
+            "off",
+        ),
+        (
+            "button.i4_edrive40_deactivate_air_conditioning",
+            "switch.i4_edrive40_climate",
+            "off",
+            "on",
+        ),
+        (
+            "button.i4_edrive40_find_vehicle",
+            "device_tracker.i4_edrive40",
+            "not_home",
+            "home",
+        ),
+    ],
+)
+async def test_service_call_success_state_change(
+    hass: HomeAssistant,
+    entity_id: str,
+    state_entity_id: str,
+    new_value: str,
+    old_value: str,
+    bmw_fixture: respx.Router,
+) -> None:
+    """Test successful button press with state change."""
+
+    # Setup component
+    assert await setup_mocked_integration(hass)
+    hass.states.async_set(state_entity_id, old_value)
+    assert hass.states.get(state_entity_id).state == old_value
+
+    # Test
+    await hass.services.async_call(
+        "button",
+        "press",
+        blocking=True,
+        target={"entity_id": entity_id},
+    )
+    check_remote_service_call(bmw_fixture)
```

Comment:
Doesn't the api call have different parameters depending on what button was pressed? Shouldn't we assert those specific parameters in the api call?

Reply:
That's true. You had another comment that we should rather test end-to-end functionality so the check if the correct button was pressed is now done in the library's API mock ([bimmer_connected/tests/common.py#L209](https://github.com/bimmerconnected/bimmer_connected/blob/master/bimmer_connected/tests/common.py#L209))

In HA, we just test if the expected change happened. If yes, then the call is assumed to be done correctly.
For services that don't change the state you're right, we don't test that explicitly, we just test if it is running without error ([bmw_connected_drive/test_button.py#L37](https://github.com/rikroe/core/blob/0803702ce1bbe3e5a3a6a6eed73ca8e5202fb5d9/tests/components/bmw_connected_drive/test_button.py#L37)). You feel we should do this?

Reply:
I think we should both test that the state in Home Assistant is correct after the service call and that the expected api call, including parameters, is done.

Reply:
👍 Will adjust. However this will need another update of the common testing code on the library.

Reply:
Ok, this will require an update to the tests in the lib (as we get the logic from there). 
The basics are there and just need some small changes. I'll also make a change that the vehicles in the mock are always ordered the same way (which should remove the need for `sorted` and ensure consistent tests across platforms).